### PR TITLE
Implement stringify (#) for preprocessor macros

### PR DESCRIPTION
**Pull request by tnibler**
Stumbled upon #2407, so I though I'd give it a shot. 

I tried touching as little as possible so nothing should break hopefully.

The rules sort of follow what GCC does, barring the most complex stuff that would probably need a complete restructuring to implement. A notable difference is that [this indirection](https://stackoverflow.com/a/2653351) isn't needed, arguments are expanded and stringified in one step. That's because the actual stringifying is done higher up in the preprocessing loop with marker tokens, which avoids having to thread more state through the entire machinery like for pasting. I feel that's very error prone (in fact a `currentPos = savePos` was missing in the pasting logic AFAICT) for not much benefit here. 

Draft because:
 - One `tokenPaste.vert` test broke, something up with recursive macros or something
 - Nested stringification is not handled, they are just flattened into one string. The nesting depth is available, so nothing stops us from building the quote/backslash escapes. Just wasn't sure if you'd actually want that, because ad-hoc escaping is (almost) always incorrect and this is just a debug feature so whatever.
 - No tests that work within the current setup, because it's tricky to test all the edge cases and produce legal code at the same time ;)

But this is what it produces:
```
#version 450
#extension GL_EXT_debug_printf : enable

#define PASTE(arg) pasted_##arg
#define STRINGIFY(arg) #arg
#define STRINGIFY_TWICE(arg, regular) before #arg #arg after regular
#define STRINGIFY_WITH_SPACE(arg) before # arg
#define PASTE_WITH_SPACE(arg) pasted_## arg
#define BOTH(arg1,    arg2) pasted_##arg1 and     #arg2
#define NOARGS(arg) 1 1 1 1
#define REGULAR(arg1, arg2) arg1    : arg2 :arg1
#define STR(a) STRHELP(a)
#define STRHELP(a) #a // this isn't necessary right now, as would be in C
#define LEGAL(arg) a######arg

void main() {
    REGULAR(8,9);
    STRINGIFY(12 3);
    STRINGIFY_TWICE(letters_too,       1);
    STRINGIFY_WITH_SPACE(aaa1);
    PASTE(symbol);
    PASTE_WITH_SPACE(symbol);
    BOTH(symbol, string);
    STRHELP(REGULAR(3, 4));
    STR(REGULAR(3, 4));
    STR(aaa     );
    STR(STR(hello));

```
```
#version 450
#extension GL_EXT_debug_printf : enable
[...]
void main() {
               8 : 9 : 8;
                  "12 3";
                                        before "letters_too" "letters_too" after 1;
                             before "aaa1";
                pasted_symbol;
                           pasted_symbol;
                       pasted_symbol and "string";
                         "3 : 4 :3";
                     "3 : 4 :3";
                "aaa";
                  "hello";
}
```

### CODE REVIEW

File: glslang/MachineIndependent/preprocessor/Pp.cpp [hunk 1]
```diff
@@ -1119,7 +1164,7 @@ int TPpContext::tMacroInput::scan(TPpToken* ppToken)
     }
 
     // see if are preceding a ##
-    if (mac->body.peekUntokenizedPasting()) {
```
Comment (tnibler) [inline glslang/MachineIndependent/preprocessor/Pp.cpp:1122]: UntokenizedPasting doesn't exist anymore, because all `#` handling is now already done at macro parsing time.

File: glslang/MachineIndependent/preprocessor/PpTokens.cpp [hunk 2]
```diff
@@ -146,8 +136,10 @@ bool TPpContext::TokenStream::peekTokenizedPasting(bool lastTokenPastes)
 
     // 2. last token and we've been told after this there will be a ##
 
-    if (! lastTokenPastes)
+    if (! lastTokenPastes) {
```
Comment (tnibler) [inline glslang/MachineIndependent/preprocessor/PpTokens.cpp:139]: I think this was missing, independently of the new feature
